### ISSUE TITLE
About etcd clientv3 non-blocking distributed locks

### ISSUE THREAD

**Issue by dglei** [etcd-io/etcd#10493]
File: [etcd/clientv3/concurrency/mutex.go#L42](https://github.com/etcd-io/etcd/blob/master/clientv3/concurrency/mutex.go#L42)
func (*Mutex).Lock(ctx Context) error {}

This function always blocks when it requests a lock that has been occupied. How can i make it non-blocking return with error?

**Comment by jingyih** [comment]
As described in the function comment, you can pass a context with timeout when calling Lock(). Upon timeout, the context is canceled, and Lock() returns with error.
https://github.com/etcd-io/etcd/blob/8c228d692bfa516e1c0977922f061b9a0fb1ae0f/clientv3/concurrency/mutex.go#L40-L41

**Comment by dglei** [comment]
> As described in the function comment, you can pass a context with timeout when calling Lock(). Upon timeout, the context is canceled, and Lock() returns with error.
> [etcd/clientv3/concurrency/mutex.go](https://github.com/etcd-io/etcd/blob/8c228d692bfa516e1c0977922f061b9a0fb1ae0f/clientv3/concurrency/mutex.go#L40-L41)
> 
> Lines 40 to 41 in [8c228d6](/etcd-io/etcd/commit/8c228d692bfa516e1c0977922f061b9a0fb1ae0f)
> 
>  // Lock locks the mutex with a cancelable context. If the context is canceled 
>  // while trying to acquire the lock, the mutex tries to clean its stale lock entry.

@jingyih 
I means if the lock is occupied when trylock, it will return immediately, instead of waitDeletes.  Passing a context with a timeout when calling Lock() does not return immediately if the lock is occupied. How long is the timeout setting appropriate?

If we need to do one thing and only do it once, it will take some time to do this, such as 1ms. The context timeout passed in lock is set to 2ms. When a goroutine gets the lock, the lock is released after 1ms, and the other coroutine can still obtain the lock.

**Comment by jingyih** [comment]
I see what you mean. I think 'tryLock' is a nice to have function. In the meantime, for your use case, can you try not to hold the lock while processing the event. Instead, use a flag to record if the event is being processed or not. And use the lock to protect this flag.

**Comment by dglei** [comment]
@jingyih @technoweenie @bmizerany @drnic This feature will not be added?

**Comment by jingyih** [comment]
Personally I am not aware of any ongoing effort on this. You are welcome to add this feature if you are interested:)

**Comment by leonchen83** [comment]
`tryLock`  feature is a good idea. we are facing the same issue.
We write a service that can switch `master` and `slave` via etcd lock
and do not want to block the service. so we also need a `tryLock` method to period check.
we found that when lease was expired, the lock method also blocked. 
pseudo code like following
```
Future<id> future = lease.grant(15);
long leaseId = future.get();
Future<String> lockFuture = lock.lock("some lock", leaseId);
lockFuture.get(); // blocking here, when lease is expired, also blocking here
// lockFuture.get(5, TimeUnit.SECONDS); this will not block invoke, but the server also runing lock method
```

**Comment by vsbogd** [comment]
@dglei, I have read your case and my understanding is that you have two go-routines which are trying to use lock as flag that some event is processed. But for event processing it should be more natural to implement the following in each thread:
```
acquire lock
get event from queue
release lock
```
Then second thread after acquiring lock will find that queue is empty and event will be processed only once.
Could you please explain your case further?
